File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/timed/DisableInactiveAuthUsersService.java
package uk.gov.justice.digital.hmpps.oauth2server.timed;

import com.microsoft.applicationinsights.TelemetryClient;
import lombok.extern.log4j.Log4j2;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository;

import java.time.LocalDateTime;
import java.util.Map;

@Service
@Log4j2
public class DisableInactiveAuthUsersService implements BatchUserService {
    private final UserRepository repository;
    private final TelemetryClient telemetryClient;
    private final int ageTrigger;

    public DisableInactiveAuthUsersService(final UserRepository repository,
                                           final TelemetryClient telemetryClient,
                                           @Value("${application.authentication.disable.age-trigger}") final int ageTrigger) {
        this.repository = repository;
        this.telemetryClient = telemetryClient;
        this.ageTrigger = ageTrigger;
    }

    @Transactional(transactionManager = "authTransactionManager")
    public int processInBatches() {
        final var usersToDisable = repository.findTop10ByLastLoggedInBeforeAndEnabledIsTrueAndMasterIsTrueOrderByLastLoggedIn(LocalDateTime.now().minusDays(ageTrigger));
        usersToDisable.forEach(user -> {
            user.setEnabled(false);
            log.debug("Disabling auth user {} due to inactivity", user.getUsername());
            telemetryClient.trackEvent("DisableInactiveAuthUsersProcessed", Map.of("username", user.getUsername()), null);
        });
        return usersToDisable.size();
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/timed/DisableInactiveAuthUsersService.kt
package uk.gov.justice.digital.hmpps.oauth2server.timed

import com.microsoft.applicationinsights.TelemetryClient
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository
import java.time.LocalDateTime
import java.util.function.Consumer

@Service
class DisableInactiveAuthUsersService(
  private val repository: UserRepository,
  private val telemetryClient: TelemetryClient,
  @Value("\${application.authentication.disable.login-days}") private val loginDays: Int,
) : BatchUserService {

  companion object {
    private val log = LoggerFactory.getLogger(this::class.java)
  }

  @Transactional(transactionManager = "authTransactionManager")
  override fun processInBatches(): Int {
    val usersToDisable = repository.findTop10ByLastLoggedInBeforeAndEnabledIsTrueAndMasterIsTrueOrderByLastLoggedIn(
      LocalDateTime.now().minusDays(loginDays.toLong())
    )
    usersToDisable.forEach(
      Consumer { user: User ->
        user.isEnabled = false
        user.inactiveReason = "INACTIVE FOR 90 DAYS"
        log.debug("Disabling auth user {} due to inactivity", user.username)
        telemetryClient.trackEvent("DisableInactiveAuthUsersProcessed", mapOf("username" to user.username), null)
      }
    )
    return usersToDisable.size
  }
}
